vimeo/VimeoNetworking
# VIM-4851

#### Ticket
[VIM-4851](https://vimean.atlassian.net/browse/VIM-4851)
Related to: https://github.vimeows.com/MobileApps/Vimeo-iOS/pull/1162

#### Ticket Summary
The unresponsive blank screen occurred when the client token became invalid. When an invalid token was used, `VimeoNetworking` then issued an `invalidToken` notification. The `invalidToken` handler method then tried to authenticate via the `authenticateAndPresentInitialInterfaceWithOnboarding` method, in which the `AuthenticationController` class kept grabbing the existing keychain that was associated with the invalid token. That flow created an infinite loop that caused the app to "hang".

#### Implementation Summary
My solution is to grant client credential right inside the `invalidToken` handler method. In addition, to prevent the handler to be called multiple times before the response comes back from the server, I check whether the token sent along the notification and the one in `VimeoClient` are not the same (for the case of user credential) and whether the `currentAccount` property of `VimeoClient` is invalid (for the case of client credential).

#### How to Test
1. Open the app
2. To simulate a situation where a token is invalidated, use breakpoints and `lldb` debugger to obtain the token, then use ModWorlds to invalidate the token.
3. Notice that the app can still recover from the token invalidation.
